# Prework 1
## Python for DataFrames

In this chapter, we review the fundamentals of Python for Dataframes, including the main methods for:

 - Import libraries and new data
 - How our data is
 - Filter depending on conditions
 - create, drop and rename columns
 - drop duplicates and sort values
 - join dataframes

**At the end of the notebook you have 3 exercises to complete**

### Importing libraries

The most important libraries we would probably need during data management

In [1]:
import pandas as pd # for data importing and DataFrame methods
import numpy as np #for working DataFrames and arrays
import matplotlib as plt # for plots and charts
import seaborn as sns # for plots

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


#### Import data

There exist multiple dataset formats, the most common: txt, csv, excel and json

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_json.html

In [3]:
df = pd.read_csv('/content/drive/MyDrive/USC Upstate/IronHack/Prework/DATA/data/census.csv', sep = ';')

# df = pd.read_excel('..\Advanced DA\data\census.xlsx')

df.head(5) # showing the first 5 raws

CensusId    State   County  TotalPop    Men  Women  Hispanic  White  Black  \
0      1001  Alabama  Autauga     55221  26745  28476       2.6   75.8   18.5   
1      1003  Alabama  Baldwin    195121  95314  99807       4.5   83.1    9.5   
2      1005  Alabama  Barbour     26932  14497  12435       4.6   46.2   46.7   
3      1007  Alabama     Bibb     22604  12073  10531       2.2   74.5   21.4   
4      1009  Alabama   Blount     57710  28512  29198       8.6   87.9    1.5   

   Native  ...  Walk  OtherTransp  WorkAtHome  MeanCommute  Employed  \
0     0.4  ...   0.5          1.3         1.8         26.5     23986   
1     0.6  ...   1.0          1.4         3.9         26.4     85953   
2     0.2  ...   1.8          1.5         1.6         24.1      8597   
3     0.4  ...   0.6          1.5         0.7         28.8      8294   
4     0.3  ...   0.9          0.4         2.3         34.9     22189   

   PrivateWork  PublicWork  SelfEmployed  FamilyWork  Unemployment  
0         73.6        20.9           5.5         0.0           7.6  
1         81.5        12.3           5.8         0.4           7.5  
2         71.8        20.8           7.3         0.1          17.6  
3         76.8        16.1           6.7         0.4           8.3  
4         82.0        13.5           4.2         0.4           7.7  

[5 rows x 37 columns]

#### Defining our own data

We can as well create our dataset

In [4]:
familia = {'miembro':('padre','madre','hijo','hija','sobrino','primo','abuela'),
           'nombre':['Manuel','Maria','Diego','Bernarda','Pablo','Antonio','Concha'],
           'edad': [58, 56, 17, 23, 5, 12, 78],
           'estado_civil':['casado','casado','soltero','soltero','soltero','soltero','viuda'],
           'altura':[1.78, 1.69, 1.7, 1.79, 1.56, 1.59, 1.60],
           'sexo':['hombro','mujer','hombre','mujer','hombre','hombre','mujer'],
           'hombre':[True, False, True, False, True, True, False]}

familia_df = pd.DataFrame(familia)
familia_df

miembro    nombre  edad estado_civil  altura    sexo  hombre
0    padre    Manuel    58       casado    1.78  hombro    True
1    madre     Maria    56       casado    1.69   mujer   False
2     hijo     Diego    17      soltero    1.70  hombre    True
3     hija  Bernarda    23      soltero    1.79   mujer   False
4  sobrino     Pablo     5      soltero    1.56  hombre    True
5    primo   Antonio    12      soltero    1.59  hombre    True
6   abuela    Concha    78        viuda    1.60   mujer   False

### How our data is

This section highlights two methods to analyze the shape of our dataframe and format of our variables

`df.shape` returns a tuple with two values, the number of raws and the number of columns

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.shape.html

`df.info()` returns the list of variables we got with the number of non-null values and the data type (integer, string, datetime...)
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html


In [5]:
df.shape

(3220, 37)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3220 entries, 0 to 3219
Data columns (total 37 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CensusId         3220 non-null   int64  
 1   State            3220 non-null   object 
 2   County           3220 non-null   object 
 3   TotalPop         3220 non-null   int64  
 4   Men              3220 non-null   int64  
 5   Women            3220 non-null   int64  
 6   Hispanic         3220 non-null   float64
 7   White            3220 non-null   float64
 8   Black            3220 non-null   float64
 9   Native           3220 non-null   float64
 10  Asian            3220 non-null   float64
 11  Pacific          3220 non-null   float64
 12  Citizen          3220 non-null   int64  
 13  Income           3219 non-null   float64
 14  IncomeErr        3219 non-null   float64
 15  IncomePerCap     3220 non-null   int64  
 16  IncomePerCapErr  3220 non-null   int64  
 17  Poverty       

### Filtering data

Filter data consists on retaining the instances/observations/raw data that meet one or several conditions, for example in the dataset above, The states with a White percentage higher than 70

https://www.geeksforgeeks.org/ways-to-filter-pandas-dataframe-by-column-values/


In [7]:
df_whites = df[(df['White'] > 60) & (df['Hispanic'] <= 6)]
df_whites

CensusId    State    County  TotalPop    Men  Women  Hispanic  White  \
0         1001  Alabama   Autauga     55221  26745  28476       2.6   75.8   
1         1003  Alabama   Baldwin    195121  95314  99807       4.5   83.1   
3         1007  Alabama      Bibb     22604  12073  10531       2.2   74.5   
7         1015  Alabama   Calhoun    116648  56274  60374       3.5   73.0   
9         1019  Alabama  Cherokee     26008  12975  13033       1.5   91.7   
...        ...      ...       ...       ...    ...    ...       ...    ...   
3130     56023  Wyoming   Lincoln     18316   9356   8960       4.4   92.8   
3132     56027  Wyoming  Niobrara      2516   1164   1352       2.6   91.5   
3133     56029  Wyoming      Park     28985  14230  14755       5.8   90.5   
3135     56033  Wyoming  Sheridan     29738  14695  15043       4.2   91.7   
3141     56045  Wyoming    Weston      7152   3695   3457       3.3   92.2   

      Black  Native  ...  Walk  OtherTransp  WorkAtHome  MeanCommute  \
0      18.5     0.4  ...   0.5          1.3         1.8         26.5   
1       9.5     0.6  ...   1.0          1.4         3.9         26.4   
3      21.4     0.4  ...   0.6          1.5         0.7         28.8   
7      20.3     0.2  ...   1.2          1.2         2.7         24.1   
9       4.8     0.6  ...   0.6          0.7         2.5         27.4   
...     ...     ...  ...   ...          ...         ...          ...   
3130    0.1     0.3  ...   3.6          0.5         7.1         29.9   
3132    0.3     1.7  ...  12.1          2.8        11.6         17.2   
3133    0.8     0.5  ...   5.6          1.5         6.3         17.2   
3135    0.8     1.1  ...   3.6          1.7         5.5         16.2   
3141    0.2     0.7  ...   3.4          1.2         5.5         29.1   

      Employed  PrivateWork  PublicWork  SelfEmployed  FamilyWork  \
0        23986         73.6        20.9           5.5         0.0   
1        85953         81.5        12.3           5.8         0.4   
3         8294         76.8        16.1           6.7         0.4   
7        47401         74.1        20.8           5.0         0.1   
9        10155         73.1        18.5           7.9         0.5   
...        ...          ...         ...           ...         ...   
3130      8724         73.6        19.8           6.2         0.3   
3132      1114         42.2        38.0          18.7         1.2   
3133     15305         75.7        16.9           7.3         0.1   
3135     14336         70.7        21.0           7.7         0.5   
3141      3343         70.2        20.2           8.3         1.3   

      Unemployment  
0              7.6  
1              7.5  
3              8.3  
7             12.3  
9              7.9  
...            ...  
3130           6.4  
3132           1.6  
3133           3.8  
3135           3.4  
3141           5.2  

[1809 rows x 37 columns]

#### Keep columns and Drop columns

Columns selection is a very common action when working with DataFrames



`.columns` method returns a list with the column names

In [8]:
df.columns

Index(['CensusId', 'State', 'County', 'TotalPop', 'Men', 'Women', 'Hispanic',
       'White', 'Black', 'Native', 'Asian', 'Pacific', 'Citizen', 'Income',
       'IncomeErr', 'IncomePerCap', 'IncomePerCapErr', 'Poverty',
       'ChildPoverty', 'Professional', 'Service', 'Office', 'Construction',
       'Production', 'Drive', 'Carpool', 'Transit', 'Walk', 'OtherTransp',
       'WorkAtHome', 'MeanCommute', 'Employed', 'PrivateWork', 'PublicWork',
       'SelfEmployed', 'FamilyWork', 'Unemployment'],
      dtype='object')

#### Keep columns

We keep State, County and TotalPop variables

In [9]:
df_keep = df[['State','County','TotalPop']]
df_keep.head()

State   County  TotalPop
0  Alabama  Autauga     55221
1  Alabama  Baldwin    195121
2  Alabama  Barbour     26932
3  Alabama     Bibb     22604
4  Alabama   Blount     57710

In [10]:
df_keep.columns

Index(['State', 'County', 'TotalPop'], dtype='object')

#### Drop columns

After passing the drop method State, County and TotalPop variables are deleted

In [11]:
df_drop = df.drop(['State','County','TotalPop'], axis = 1)
df_drop.head()

CensusId    Men  Women  Hispanic  White  Black  Native  Asian  Pacific  \
0      1001  26745  28476       2.6   75.8   18.5     0.4    1.0      0.0   
1      1003  95314  99807       4.5   83.1    9.5     0.6    0.7      0.0   
2      1005  14497  12435       4.6   46.2   46.7     0.2    0.4      0.0   
3      1007  12073  10531       2.2   74.5   21.4     0.4    0.1      0.0   
4      1009  28512  29198       8.6   87.9    1.5     0.3    0.1      0.0   

   Citizen  ...  Walk  OtherTransp  WorkAtHome  MeanCommute  Employed  \
0    40725  ...   0.5          1.3         1.8         26.5     23986   
1   147695  ...   1.0          1.4         3.9         26.4     85953   
2    20714  ...   1.8          1.5         1.6         24.1      8597   
3    17495  ...   0.6          1.5         0.7         28.8      8294   
4    42345  ...   0.9          0.4         2.3         34.9     22189   

   PrivateWork  PublicWork  SelfEmployed  FamilyWork  Unemployment  
0         73.6        20.9           5.5         0.0           7.6  
1         81.5        12.3           5.8         0.4           7.5  
2         71.8        20.8           7.3         0.1          17.6  
3         76.8        16.1           6.7         0.4           8.3  
4         82.0        13.5           4.2         0.4           7.7  

[5 rows x 34 columns]

In [12]:
df_drop.columns

Index(['CensusId', 'Men', 'Women', 'Hispanic', 'White', 'Black', 'Native',
       'Asian', 'Pacific', 'Citizen', 'Income', 'IncomeErr', 'IncomePerCap',
       'IncomePerCapErr', 'Poverty', 'ChildPoverty', 'Professional', 'Service',
       'Office', 'Construction', 'Production', 'Drive', 'Carpool', 'Transit',
       'Walk', 'OtherTransp', 'WorkAtHome', 'MeanCommute', 'Employed',
       'PrivateWork', 'PublicWork', 'SelfEmployed', 'FamilyWork',
       'Unemployment'],
      dtype='object')

#### Locate inside a DataFrame


There two methods which allow us to Slice and Dice our dataframe: `.loc` and `.iloc`

`.iloc` Access the rows and columns by index

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html

In this first example we keep the first 15 rows

In [13]:
df.iloc[0:15]

CensusId    State    County  TotalPop    Men  Women  Hispanic  White  \
0       1001  Alabama   Autauga     55221  26745  28476       2.6   75.8   
1       1003  Alabama   Baldwin    195121  95314  99807       4.5   83.1   
2       1005  Alabama   Barbour     26932  14497  12435       4.6   46.2   
3       1007  Alabama      Bibb     22604  12073  10531       2.2   74.5   
4       1009  Alabama    Blount     57710  28512  29198       8.6   87.9   
5       1011  Alabama   Bullock     10678   5660   5018       4.4   22.2   
6       1013  Alabama    Butler     20354   9502  10852       1.2   53.3   
7       1015  Alabama   Calhoun    116648  56274  60374       3.5   73.0   
8       1017  Alabama  Chambers     34079  16258  17821       0.4   57.3   
9       1019  Alabama  Cherokee     26008  12975  13033       1.5   91.7   
10      1021  Alabama   Chilton     43819  21619  22200       7.6   80.5   
11      1023  Alabama   Choctaw     13395   6382   7013       0.4   55.9   
12      1025  Alabama    Clarke     25070  11834  13236       0.3   53.4   
13      1027  Alabama      Clay     13537   6671   6866       3.2   79.9   
14      1029  Alabama  Cleburne     15002   7334   7668       2.3   92.5   

    Black  Native  ...  Walk  OtherTransp  WorkAtHome  MeanCommute  Employed  \
0    18.5     0.4  ...   0.5          1.3         1.8         26.5     23986   
1     9.5     0.6  ...   1.0          1.4         3.9         26.4     85953   
2    46.7     0.2  ...   1.8          1.5         1.6         24.1      8597   
3    21.4     0.4  ...   0.6          1.5         0.7         28.8      8294   
4     1.5     0.3  ...   0.9          0.4         2.3         34.9     22189   
5    70.7     1.2  ...   5.0          1.7         2.8         27.5      3865   
6    43.8     0.1  ...   0.8          0.6         1.7         24.6      7813   
7    20.3     0.2  ...   1.2          1.2         2.7         24.1     47401   
8    40.3     0.2  ...   0.3          0.4         2.1         25.1     13689   
9     4.8     0.6  ...   0.6          0.7         2.5         27.4     10155   
10   10.2     0.4  ...   1.1          1.4         1.9         32.0     17895   
11   42.9     0.0  ...   1.9          1.4         0.8         30.7      4405   
12   45.3     0.0  ...   1.3          2.6         1.1         23.0      8161   
13   14.4     0.7  ...   1.8          0.9         2.1         30.3      5180   
14    2.9     0.2  ...   0.2          1.0         3.4         33.3      6065   

    PrivateWork  PublicWork  SelfEmployed  FamilyWork  Unemployment  
0          73.6        20.9           5.5         0.0           7.6  
1          81.5        12.3           5.8         0.4           7.5  
2          71.8        20.8           7.3         0.1          17.6  
3          76.8        16.1           6.7         0.4           8.3  
4          82.0        13.5           4.2         0.4           7.7  
5          79.5        15.1           5.4         0.0          18.0  
6          77.4        16.2           6.2         0.2          10.9  
7          74.1        20.8           5.0         0.1          12.3  
8          85.1        12.1           2.8         0.0           8.9  
9          73.1        18.5           7.9         0.5           7.9  
10         80.3        15.2           4.1         0.5           9.1  
11         77.5        16.8           5.4         0.3          13.6  
12         83.1        13.6           3.3         0.0          19.4  
13         77.5        14.7           7.8         0.0           9.4  
14         76.3        15.3           8.4         0.0           8.3  

[15 rows x 37 columns]

Next, we keep only from the second row to the 13th, and columns 10 to 15

In [14]:
df.iloc[2:14,10:15]

Asian  Pacific  Citizen   Income  IncomeErr
2     0.4      0.0    20714  32964.0     2973.0
3     0.1      0.0    17495  38678.0     3995.0
4     0.1      0.0    42345  45813.0     3141.0
5     0.2      0.0     8057  31938.0     5884.0
6     0.4      0.0    15581  32229.0     1793.0
7     0.9      0.0    88612  41703.0      925.0
8     0.8      0.0    26462  34177.0     2949.0
9     0.3      0.0    20600  36296.0     1710.0
10    0.3      0.0    31728  41627.0     2025.0
11    0.0      0.0    10568  33536.0     2231.0
12    0.4      0.4    19258  32011.0     3088.0
13    0.0      0.0    10312  35327.0     4517.0

`.loc` Access a group of rows and columns by label(s) or a boolean array

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html

Analogously to filtering (above) we can use `.loc`. Later we will see exactly in which use cases `.loc` is more useful

In [15]:
df_whites = df.loc[(df['White'] > 60) & (df['Hispanic'] <= 6)]
df_whites

CensusId    State    County  TotalPop    Men  Women  Hispanic  White  \
0         1001  Alabama   Autauga     55221  26745  28476       2.6   75.8   
1         1003  Alabama   Baldwin    195121  95314  99807       4.5   83.1   
3         1007  Alabama      Bibb     22604  12073  10531       2.2   74.5   
7         1015  Alabama   Calhoun    116648  56274  60374       3.5   73.0   
9         1019  Alabama  Cherokee     26008  12975  13033       1.5   91.7   
...        ...      ...       ...       ...    ...    ...       ...    ...   
3130     56023  Wyoming   Lincoln     18316   9356   8960       4.4   92.8   
3132     56027  Wyoming  Niobrara      2516   1164   1352       2.6   91.5   
3133     56029  Wyoming      Park     28985  14230  14755       5.8   90.5   
3135     56033  Wyoming  Sheridan     29738  14695  15043       4.2   91.7   
3141     56045  Wyoming    Weston      7152   3695   3457       3.3   92.2   

      Black  Native  ...  Walk  OtherTransp  WorkAtHome  MeanCommute  \
0      18.5     0.4  ...   0.5          1.3         1.8         26.5   
1       9.5     0.6  ...   1.0          1.4         3.9         26.4   
3      21.4     0.4  ...   0.6          1.5         0.7         28.8   
7      20.3     0.2  ...   1.2          1.2         2.7         24.1   
9       4.8     0.6  ...   0.6          0.7         2.5         27.4   
...     ...     ...  ...   ...          ...         ...          ...   
3130    0.1     0.3  ...   3.6          0.5         7.1         29.9   
3132    0.3     1.7  ...  12.1          2.8        11.6         17.2   
3133    0.8     0.5  ...   5.6          1.5         6.3         17.2   
3135    0.8     1.1  ...   3.6          1.7         5.5         16.2   
3141    0.2     0.7  ...   3.4          1.2         5.5         29.1   

      Employed  PrivateWork  PublicWork  SelfEmployed  FamilyWork  \
0        23986         73.6        20.9           5.5         0.0   
1        85953         81.5        12.3           5.8         0.4   
3         8294         76.8        16.1           6.7         0.4   
7        47401         74.1        20.8           5.0         0.1   
9        10155         73.1        18.5           7.9         0.5   
...        ...          ...         ...           ...         ...   
3130      8724         73.6        19.8           6.2         0.3   
3132      1114         42.2        38.0          18.7         1.2   
3133     15305         75.7        16.9           7.3         0.1   
3135     14336         70.7        21.0           7.7         0.5   
3141      3343         70.2        20.2           8.3         1.3   

      Unemployment  
0              7.6  
1              7.5  
3              8.3  
7             12.3  
9              7.9  
...            ...  
3130           6.4  
3132           1.6  
3133           3.8  
3135           3.4  
3141           5.2  

[1809 rows x 37 columns]

### Create columns

Sometimes it is very interesitng to to create new columns based on existing ones. Here are a few examples of how to accomplish this task.

Here, `.loc` is one of the most common ways to define new columns, let's see how

We create a new variable called `Unemployment_rate` , if `Unemployment` is higher than 10 we assign the value `High` to the new column, contrarily, if lower than 10 the new variable takes the value `Low`

In [16]:
df.loc[df['Unemployment'] > 10, 'Unemployment_rate'] = 'High'
df.loc[df['Unemployment'] <= 10, 'Unemployment_rate'] = 'Low'
df

CensusId        State     County  TotalPop    Men  Women  Hispanic  \
0         1001      Alabama    Autauga     55221  26745  28476       2.6   
1         1003      Alabama    Baldwin    195121  95314  99807       4.5   
2         1005      Alabama    Barbour     26932  14497  12435       4.6   
3         1007      Alabama       Bibb     22604  12073  10531       2.2   
4         1009      Alabama     Blount     57710  28512  29198       8.6   
...        ...          ...        ...       ...    ...    ...       ...   
3215     72145  Puerto Rico  Vega Baja     56858  27379  29479      96.4   
3216     72147  Puerto Rico    Vieques      9130   4585   4545      96.7   
3217     72149  Puerto Rico   Villalba     24685  12086  12599      99.7   
3218     72151  Puerto Rico    Yabucoa     36279  17648  18631      99.8   
3219     72153  Puerto Rico      Yauco     39474  19047  20427      99.5   

      White  Black  Native  ...  OtherTransp  WorkAtHome  MeanCommute  \
0      75.8   18.5     0.4  ...          1.3         1.8         26.5   
1      83.1    9.5     0.6  ...          1.4         3.9         26.4   
2      46.2   46.7     0.2  ...          1.5         1.6         24.1   
3      74.5   21.4     0.4  ...          1.5         0.7         28.8   
4      87.9    1.5     0.3  ...          0.4         2.3         34.9   
...     ...    ...     ...  ...          ...         ...          ...   
3215    3.4    0.1     0.0  ...          1.3         0.3         32.0   
3216    2.9    0.0     0.0  ...          0.0         1.4         14.0   
3217    0.0    0.0     0.0  ...          0.0         3.3         26.9   
3218    0.2    0.0     0.0  ...          2.3         1.5         29.5   
3219    0.5    0.0     0.0  ...          0.7         3.1         24.6   

      Employed  PrivateWork  PublicWork  SelfEmployed  FamilyWork  \
0        23986         73.6        20.9           5.5         0.0   
1        85953         81.5        12.3           5.8         0.4   
2         8597         71.8        20.8           7.3         0.1   
3         8294         76.8        16.1           6.7         0.4   
4        22189         82.0        13.5           4.2         0.4   
...        ...          ...         ...           ...         ...   
3215     13660         78.3        17.6           4.1         0.0   
3216      2860         44.5        41.6          13.6         0.3   
3217      6795         59.2        27.5          13.1         0.2   
3218      8083         65.1        27.6           7.3         0.0   
3219      8923         68.0        27.6           4.4         0.0   

      Unemployment  Unemployment_rate  
0              7.6                Low  
1              7.5                Low  
2             17.6               High  
3              8.3                Low  
4              7.7                Low  
...            ...                ...  
3215          15.2               High  
3216          12.2               High  
3217          25.9               High  
3218          24.3               High  
3219          27.1               High  

[3220 rows x 38 columns]

Next we have an alternative way of creating a new variable from an existing one. Firstly, we define a function, then we call it through `apply` method. Be sure you save the transformation in a new variable!

Both ways are similar

In [17]:
def Unemployment_new_var(x):
    if x > 10:
        return 'High'
    else:
        return 'Low'

df['Unemployment_rate'] = df['Unemployment'].apply(Unemployment_new_var)
df

CensusId        State     County  TotalPop    Men  Women  Hispanic  \
0         1001      Alabama    Autauga     55221  26745  28476       2.6   
1         1003      Alabama    Baldwin    195121  95314  99807       4.5   
2         1005      Alabama    Barbour     26932  14497  12435       4.6   
3         1007      Alabama       Bibb     22604  12073  10531       2.2   
4         1009      Alabama     Blount     57710  28512  29198       8.6   
...        ...          ...        ...       ...    ...    ...       ...   
3215     72145  Puerto Rico  Vega Baja     56858  27379  29479      96.4   
3216     72147  Puerto Rico    Vieques      9130   4585   4545      96.7   
3217     72149  Puerto Rico   Villalba     24685  12086  12599      99.7   
3218     72151  Puerto Rico    Yabucoa     36279  17648  18631      99.8   
3219     72153  Puerto Rico      Yauco     39474  19047  20427      99.5   

      White  Black  Native  ...  OtherTransp  WorkAtHome  MeanCommute  \
0      75.8   18.5     0.4  ...          1.3         1.8         26.5   
1      83.1    9.5     0.6  ...          1.4         3.9         26.4   
2      46.2   46.7     0.2  ...          1.5         1.6         24.1   
3      74.5   21.4     0.4  ...          1.5         0.7         28.8   
4      87.9    1.5     0.3  ...          0.4         2.3         34.9   
...     ...    ...     ...  ...          ...         ...          ...   
3215    3.4    0.1     0.0  ...          1.3         0.3         32.0   
3216    2.9    0.0     0.0  ...          0.0         1.4         14.0   
3217    0.0    0.0     0.0  ...          0.0         3.3         26.9   
3218    0.2    0.0     0.0  ...          2.3         1.5         29.5   
3219    0.5    0.0     0.0  ...          0.7         3.1         24.6   

      Employed  PrivateWork  PublicWork  SelfEmployed  FamilyWork  \
0        23986         73.6        20.9           5.5         0.0   
1        85953         81.5        12.3           5.8         0.4   
2         8597         71.8        20.8           7.3         0.1   
3         8294         76.8        16.1           6.7         0.4   
4        22189         82.0        13.5           4.2         0.4   
...        ...          ...         ...           ...         ...   
3215     13660         78.3        17.6           4.1         0.0   
3216      2860         44.5        41.6          13.6         0.3   
3217      6795         59.2        27.5          13.1         0.2   
3218      8083         65.1        27.6           7.3         0.0   
3219      8923         68.0        27.6           4.4         0.0   

      Unemployment  Unemployment_rate  
0              7.6                Low  
1              7.5                Low  
2             17.6               High  
3              8.3                Low  
4              7.7                Low  
...            ...                ...  
3215          15.2               High  
3216          12.2               High  
3217          25.9               High  
3218          24.3               High  
3219          27.1               High  

[3220 rows x 38 columns]

### Rename columns


`rename()` method allow us to name the variables as we wish. however is not the only way.

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rename.html

In [18]:
df_rename = df.rename(columns={'Men':'Hombres','Women':'Mujeres'})
df_rename

CensusId        State     County  TotalPop  Hombres  Mujeres  Hispanic  \
0         1001      Alabama    Autauga     55221    26745    28476       2.6   
1         1003      Alabama    Baldwin    195121    95314    99807       4.5   
2         1005      Alabama    Barbour     26932    14497    12435       4.6   
3         1007      Alabama       Bibb     22604    12073    10531       2.2   
4         1009      Alabama     Blount     57710    28512    29198       8.6   
...        ...          ...        ...       ...      ...      ...       ...   
3215     72145  Puerto Rico  Vega Baja     56858    27379    29479      96.4   
3216     72147  Puerto Rico    Vieques      9130     4585     4545      96.7   
3217     72149  Puerto Rico   Villalba     24685    12086    12599      99.7   
3218     72151  Puerto Rico    Yabucoa     36279    17648    18631      99.8   
3219     72153  Puerto Rico      Yauco     39474    19047    20427      99.5   

      White  Black  Native  ...  OtherTransp  WorkAtHome  MeanCommute  \
0      75.8   18.5     0.4  ...          1.3         1.8         26.5   
1      83.1    9.5     0.6  ...          1.4         3.9         26.4   
2      46.2   46.7     0.2  ...          1.5         1.6         24.1   
3      74.5   21.4     0.4  ...          1.5         0.7         28.8   
4      87.9    1.5     0.3  ...          0.4         2.3         34.9   
...     ...    ...     ...  ...          ...         ...          ...   
3215    3.4    0.1     0.0  ...          1.3         0.3         32.0   
3216    2.9    0.0     0.0  ...          0.0         1.4         14.0   
3217    0.0    0.0     0.0  ...          0.0         3.3         26.9   
3218    0.2    0.0     0.0  ...          2.3         1.5         29.5   
3219    0.5    0.0     0.0  ...          0.7         3.1         24.6   

      Employed  PrivateWork  PublicWork  SelfEmployed  FamilyWork  \
0        23986         73.6        20.9           5.5         0.0   
1        85953         81.5        12.3           5.8         0.4   
2         8597         71.8        20.8           7.3         0.1   
3         8294         76.8        16.1           6.7         0.4   
4        22189         82.0        13.5           4.2         0.4   
...        ...          ...         ...           ...         ...   
3215     13660         78.3        17.6           4.1         0.0   
3216      2860         44.5        41.6          13.6         0.3   
3217      6795         59.2        27.5          13.1         0.2   
3218      8083         65.1        27.6           7.3         0.0   
3219      8923         68.0        27.6           4.4         0.0   

      Unemployment  Unemployment_rate  
0              7.6                Low  
1              7.5                Low  
2             17.6               High  
3              8.3                Low  
4              7.7                Low  
...            ...                ...  
3215          15.2               High  
3216          12.2               High  
3217          25.9               High  
3218          24.3               High  
3219          27.1               High  

[3220 rows x 38 columns]

If we want to rename multiple columns may be this second way is better, however you have to "rename" every column

In [19]:
df.columns = ['CensusId', 'State', 'County', 'TotalPop', 'Hombres', 'Mujeres', 'Hispanic',
       'White', 'Black', 'Native', 'Asian', 'Pacific', 'Citizen', 'Income',
       'IncomeErr', 'IncomePerCap', 'IncomePerCapErr', 'Poverty',
       'ChildPoverty', 'Professional', 'Service', 'Office', 'Construction',
       'Production', 'Drive', 'Carpool', 'Transit', 'Walk', 'OtherTransp',
       'WorkAtHome', 'MeanCommute', 'Employed', 'PrivateWork', 'PublicWork',
       'SelfEmployed', 'FamilyWork', 'Unemployment', 'Unemployment_rate']
df

CensusId        State     County  TotalPop  Hombres  Mujeres  Hispanic  \
0         1001      Alabama    Autauga     55221    26745    28476       2.6   
1         1003      Alabama    Baldwin    195121    95314    99807       4.5   
2         1005      Alabama    Barbour     26932    14497    12435       4.6   
3         1007      Alabama       Bibb     22604    12073    10531       2.2   
4         1009      Alabama     Blount     57710    28512    29198       8.6   
...        ...          ...        ...       ...      ...      ...       ...   
3215     72145  Puerto Rico  Vega Baja     56858    27379    29479      96.4   
3216     72147  Puerto Rico    Vieques      9130     4585     4545      96.7   
3217     72149  Puerto Rico   Villalba     24685    12086    12599      99.7   
3218     72151  Puerto Rico    Yabucoa     36279    17648    18631      99.8   
3219     72153  Puerto Rico      Yauco     39474    19047    20427      99.5   

      White  Black  Native  ...  OtherTransp  WorkAtHome  MeanCommute  \
0      75.8   18.5     0.4  ...          1.3         1.8         26.5   
1      83.1    9.5     0.6  ...          1.4         3.9         26.4   
2      46.2   46.7     0.2  ...          1.5         1.6         24.1   
3      74.5   21.4     0.4  ...          1.5         0.7         28.8   
4      87.9    1.5     0.3  ...          0.4         2.3         34.9   
...     ...    ...     ...  ...          ...         ...          ...   
3215    3.4    0.1     0.0  ...          1.3         0.3         32.0   
3216    2.9    0.0     0.0  ...          0.0         1.4         14.0   
3217    0.0    0.0     0.0  ...          0.0         3.3         26.9   
3218    0.2    0.0     0.0  ...          2.3         1.5         29.5   
3219    0.5    0.0     0.0  ...          0.7         3.1         24.6   

      Employed  PrivateWork  PublicWork  SelfEmployed  FamilyWork  \
0        23986         73.6        20.9           5.5         0.0   
1        85953         81.5        12.3           5.8         0.4   
2         8597         71.8        20.8           7.3         0.1   
3         8294         76.8        16.1           6.7         0.4   
4        22189         82.0        13.5           4.2         0.4   
...        ...          ...         ...           ...         ...   
3215     13660         78.3        17.6           4.1         0.0   
3216      2860         44.5        41.6          13.6         0.3   
3217      6795         59.2        27.5          13.1         0.2   
3218      8083         65.1        27.6           7.3         0.0   
3219      8923         68.0        27.6           4.4         0.0   

      Unemployment  Unemployment_rate  
0              7.6                Low  
1              7.5                Low  
2             17.6               High  
3              8.3                Low  
4              7.7                Low  
...            ...                ...  
3215          15.2               High  
3216          12.2               High  
3217          25.9               High  
3218          24.3               High  
3219          27.1               High  

[3220 rows x 38 columns]

### Drop duplicates and Sort values

These two methods completely useful, but be very careful with their use, specially with drop duplicates


https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html

#### Drop duplicates
Next we drop duplicates by `State`, so we keep only one row by `State`

In [20]:
df_duplicate = df.drop_duplicates('State')
df_duplicate

CensusId                 State                  County  TotalPop  \
0         1001               Alabama                 Autauga     55221   
67        2013                Alaska  Aleutians East Borough      3304   
96        4001               Arizona                  Apache     72124   
111       5001              Arkansas                Arkansas     18731   
186       6001            California                 Alameda   1584983   
244       8001              Colorado                   Adams    471206   
308       9001           Connecticut               Fairfield    939983   
316      10001              Delaware                    Kent    169509   
319      11001  District of Columbia    District of Columbia    647484   
320      12001               Florida                 Alachua    254218   
387      13001               Georgia                 Appling     18417   
546      15001                Hawaii                  Hawaii    191482   
551      16001                 Idaho                     Ada    417501   
595      17001              Illinois                   Adams     67081   
697      18001               Indiana                   Adams     34642   
789      19001                  Iowa                   Adair      7426   
888      20001                Kansas                   Allen     13081   
993      21001              Kentucky                   Adair     18852   
1113     22001             Louisiana                  Acadia     62163   
1177     23001                 Maine            Androscoggin    107393   
1193     24001              Maryland                Allegany     73549   
1217     25001         Massachusetts              Barnstable    214766   
1231     26001              Michigan                  Alcona     10550   
1314     27001             Minnesota                  Aitkin     15839   
1401     28001           Mississippi                   Adams     31979   
1483     29001              Missouri                   Adair     25560   
1598     30001               Montana              Beaverhead      9291   
1654     31001              Nebraska                   Adams     31442   
1747     32001                Nevada               Churchill     24252   
1764     33001         New Hampshire                 Belknap     60399   
1774     34001            New Jersey                Atlantic    275376   
1795     35001            New Mexico              Bernalillo    673943   
1828     36001              New York                  Albany    307463   
1890     37001        North Carolina                Alamance    155258   
1990     38001          North Dakota                   Adams      2341   
2043     39001                  Ohio                   Adams     28229   
2131     40001              Oklahoma                   Adair     22236   
2208     41001                Oregon                   Baker     16052   
2244     42001          Pennsylvania                   Adams    101767   
2311     44001          Rhode Island                 Bristol     49176   
2316     45001        South Carolina               Abbeville     24997   
2362     46003          South Dakota                  Aurora      2729   
2428     47001             Tennessee                Anderson     75430   
2523     48001                 Texas                Anderson     57915   
2777     49001                  Utah                  Beaver      6461   
2806     50001               Vermont                 Addison     36943   
2820     51001              Virginia                Accomack     33115   
2953     53001            Washington                   Adams     19081   
2992     54001         West Virginia                 Barbour     16731   
3047     55001             Wisconsin                   Adams     20451   
3119     56001               Wyoming                  Albany     37565   
3142     72001           Puerto Rico                Adjuntas     18962   

      Hombres  Mujeres  Hispanic  White  Black  Native  ...  OtherTransp  \
0      

Next we drop duplicates by two variables

In [21]:
df_duplicate = df.drop_duplicates(['State','County'])
df_duplicate

CensusId        State     County  TotalPop  Hombres  Mujeres  Hispanic  \
0         1001      Alabama    Autauga     55221    26745    28476       2.6   
1         1003      Alabama    Baldwin    195121    95314    99807       4.5   
2         1005      Alabama    Barbour     26932    14497    12435       4.6   
3         1007      Alabama       Bibb     22604    12073    10531       2.2   
4         1009      Alabama     Blount     57710    28512    29198       8.6   
...        ...          ...        ...       ...      ...      ...       ...   
3215     72145  Puerto Rico  Vega Baja     56858    27379    29479      96.4   
3216     72147  Puerto Rico    Vieques      9130     4585     4545      96.7   
3217     72149  Puerto Rico   Villalba     24685    12086    12599      99.7   
3218     72151  Puerto Rico    Yabucoa     36279    17648    18631      99.8   
3219     72153  Puerto Rico      Yauco     39474    19047    20427      99.5   

      White  Black  Native  ...  OtherTransp  WorkAtHome  MeanCommute  \
0      75.8   18.5     0.4  ...          1.3         1.8         26.5   
1      83.1    9.5     0.6  ...          1.4         3.9         26.4   
2      46.2   46.7     0.2  ...          1.5         1.6         24.1   
3      74.5   21.4     0.4  ...          1.5         0.7         28.8   
4      87.9    1.5     0.3  ...          0.4         2.3         34.9   
...     ...    ...     ...  ...          ...         ...          ...   
3215    3.4    0.1     0.0  ...          1.3         0.3         32.0   
3216    2.9    0.0     0.0  ...          0.0         1.4         14.0   
3217    0.0    0.0     0.0  ...          0.0         3.3         26.9   
3218    0.2    0.0     0.0  ...          2.3         1.5         29.5   
3219    0.5    0.0     0.0  ...          0.7         3.1         24.6   

      Employed  PrivateWork  PublicWork  SelfEmployed  FamilyWork  \
0        23986         73.6        20.9           5.5         0.0   
1        85953         81.5        12.3           5.8         0.4   
2         8597         71.8        20.8           7.3         0.1   
3         8294         76.8        16.1           6.7         0.4   
4        22189         82.0        13.5           4.2         0.4   
...        ...          ...         ...           ...         ...   
3215     13660         78.3        17.6           4.1         0.0   
3216      2860         44.5        41.6          13.6         0.3   
3217      6795         59.2        27.5          13.1         0.2   
3218      8083         65.1        27.6           7.3         0.0   
3219      8923         68.0        27.6           4.4         0.0   

      Unemployment  Unemployment_rate  
0              7.6                Low  
1              7.5                Low  
2             17.6               High  
3              8.3                Low  
4              7.7                Low  
...            ...                ...  
3215          15.2               High  
3216          12.2               High  
3217          25.9               High  
3218          24.3               High  
3219          27.1               High  

[3220 rows x 38 columns]

#### Sort values

We order the rows by one or more variables

In [22]:
df_sort = df.sort_values('White')
df_sort

CensusId          State        County  TotalPop  Hombres  Mujeres  \
3217     72149    Puerto Rico      Villalba     24685    12086    12599   
3189     72093    Puerto Rico       Maricao      6315     3144     3171   
3151     72019    Puerto Rico  Barranquitas     29805    14695    15110   
3186     72087    Puerto Rico         Loíza     28454    13335    15119   
3164     72045    Puerto Rico       Comerío     20339    10122    10217   
...        ...            ...           ...       ...      ...      ...   
3025     54067  West Virginia      Nicholas     25930    12779    13151   
2864     51091       Virginia      Highland      2244     1003     1241   
1069     21153       Kentucky      Magoffin     12979     6515     6464   
1047     21109       Kentucky       Jackson     13357     6650     6707   
1745     31183       Nebraska       Wheeler       847      432      415   

      Hispanic  White  Black  Native  ...  OtherTransp  WorkAtHome  \
3217      99.7    0.0    0.0     0.0  ...          0.0         3.3   
3189      99.9    0.0    0.1     0.0  ...          1.3         2.6   
3151      99.9    0.0    0.0     0.0  ...          0.0         1.8   
3186      99.9    0.0    0.0     0.0  ...          0.8         0.6   
3164      99.8    0.1    0.0     0.0  ...          3.3         0.6   
...        ...    ...    ...     ...  ...          ...         ...   
3025       0.0   99.3    0.0     0.0  ...          1.9         2.0   
2864       0.0   99.5    0.0     0.4  ...          1.0         5.0   
1069       0.3   99.5    0.1     0.0  ...          0.6         2.2   
1047       0.1   99.6    0.0     0.0  ...          1.0         0.7   
1745       0.0   99.8    0.0     0.0  ...          0.6        16.6   

      MeanCommute  Employed  PrivateWork  PublicWork  SelfEmployed  \
3217         26.9      6795         59.2        27.5          13.1   
3189         22.1      1500         62.5        29.1           8.3   
3151         29.3      6235         62.0        30.0           8.0   
3186         38.3      8443         61.9        33.7           4.5   
3164         37.2      4776         63.5        28.4           8.2   
...           ...       ...          ...         ...           ...   
3025         26.9      9824         78.6        17.3           3.5   
2864         21.3      1006         61.8        21.3          16.9   
1069         31.7      3663         74.0        22.7           2.9   
1047         28.1      3993         72.9        22.1           5.0   
1745         17.5       494         59.1        16.8          22.7   

      FamilyWork  Unemployment  Unemployment_rate  
3217         0.2          25.9               High  
3189         0.0          10.7               High  
3151         0.0           6.8                Low  
3186         0.0          26.2               High  
3164         0.0          20.6               High  
...          ...           ...                ...  
3025         0.6           8.4                Low  
2864         0.0           1.1                Low  
1069         0.4          12.2               High  
1047         0.0          12.3               High  
1745         1.4           2.9                Low  

[3220 rows x 38 columns]

In [23]:
df_sort = df.sort_values(['State', 'Mujeres'], ascending = [True, False])
df_sort.head(15)

CensusId    State      County  TotalPop  Hombres  Mujeres  Hispanic  \
36      1073  Alabama   Jefferson    659026   311581   347445       3.8   
48      1097  Alabama      Mobile    414251   198216   216035       2.6   
44      1089  Alabama     Madison    346438   169422   177016       4.7   
50      1101  Alabama  Montgomery    228138   108296   119842       3.4   
58      1117  Alabama      Shelby    203530    99134   104396       5.8   
62      1125  Alabama  Tuscaloosa    200458    96781   103677       3.2   
1       1003  Alabama     Baldwin    195121    95314    99807       4.5   
40      1081  Alabama         Lee    150982    74277    76705       3.7   
51      1103  Alabama      Morgan    119786    58814    60972       7.8   
7       1015  Alabama     Calhoun    116648    56274    60374       3.5   
34      1069  Alabama     Houston    103534    49563    53971       3.2   
27      1055  Alabama      Etowah    103766    50207    53559       3.6   
38      1077  Alabama  Lauderdale     92737    44419    48318       2.4   
47      1095  Alabama    Marshall     94318    46409    47909      12.7   
41      1083  Alabama   Limestone     88805    44626    44179       5.8   

    White  Black  Native  ...  OtherTransp  WorkAtHome  MeanCommute  Employed  \
36   51.0   42.2     0.2  ...          1.0         2.5         23.6    293202   
48   58.1   35.0     0.6  ...          0.8         3.7         24.4    170900   
44   65.6   23.9     0.6  ...          1.5         3.0         20.8    163753   
50   36.9   55.8     0.2  ...          0.7         2.8         19.3     99158   
58   78.8   11.7     0.2  ...          0.8         4.9         28.5    101343   
62   63.5   30.6     0.1  ...          0.9         2.5         22.1     87695   
1    83.1    9.5     0.6  ...          1.4         3.9         26.4     85953   
40   68.2   23.0     0.1  ...          1.0         3.1         21.7     68614   
51   76.8   12.1     0.5  ...          0.6         1.6         24.1     51472   
7    73.0   20.3     0.2  ...          1.2         2.7         24.1     47401   
34   67.7   26.2     0.3  ...          1.3         2.1         21.4     44067   
27   78.5   15.4     0.3  ...          0.5         3.8         23.4     41236   
38   84.9    9.9     0.4  ...          1.0         2.2         24.6     39734   
47   82.7    1.9     0.4  ...          0.8         3.1         23.6     38140   
41   77.4   12.8     0.5  ...          0.6         2.1         26.3     37664   

    PrivateWork  PublicWork  SelfEmployed  FamilyWork  Unemployment  \
36         80.9        14.4           4.6         0.2           9.1   
48         81.4        13.8           4.6         0.1           9.8   
44         75.6        19.5           4.7         0.2           8.5   
50         72.6        22.7           4.6         0.1           8.8   
58         82.6        12.5           4.8         0.1           5.5   
62         77.0        18.9           4.0         0.1           7.6   
1          81.5        12.3           5.8         0.4           7.5   
40         75.5        20.3           4.1         0.1           7.3   
51         81.9        12.1           5.9         0.1           9.9   
7          74.1        20.8           5.0         0.1          12.3   
34         81.3        13.0           5.5         0.2           8.3   
27         79.8        14.4           5.6         0.2           9.0   
38         77.3        16.7           5.8         0.2           7.1   
47         78.0        14.8           7.1         0.1           8.4   
41         77.9        16.4           5.5         0.2           7.8   

    Unemployment_rate  
36                Low  
48                Low  
44                Low  
50                Low  
58                Low  
62                Low  
1                 Low  
40                Low  
51                Low  
7                High  
34                Low  
27                Low  
38                Low  
47                Low  
41       

### Join dataframes


There are two methods for joining datasets  `merge` and `concat`

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html

https://www.geeksforgeeks.org/pandas-concat-function-in-python/

https://realpython.com/pandas-merge-join-and-concat/

Firstly we define two dataframes

In [81]:
familia = {'miembro':('padre','madre','hijo','hija','sobrino','primo','abuela'),
           'nombre':['Manuel','Maria','Diego','Bernarda','Pablo','Antonio','Concha'],
           'edad': [58, 56, 17, 23, 5, 12, 78],
           'estado_civil':['casado','casado','soltero','soltero','soltero','soltero','viuda'],
           'altura':[1.78, 1.69, 1.7, 1.79, 1.56, 1.59, 1.60],
           'sexo':['hombro','mujer','hombre','mujer','hombre','hombre','mujer'],
           'hombre':[True, False, True, False, True, True, False]}

familia_df = pd.DataFrame(familia)
familia_df

miembro    nombre  edad estado_civil  altura    sexo  hombre
0    padre    Manuel    58       casado    1.78  hombro    True
1    madre     Maria    56       casado    1.69   mujer   False
2     hijo     Diego    17      soltero    1.70  hombre    True
3     hija  Bernarda    23      soltero    1.79   mujer   False
4  sobrino     Pablo     5      soltero    1.56  hombre    True
5    primo   Antonio    12      soltero    1.59  hombre    True
6   abuela    Concha    78        viuda    1.60   mujer   False

In [82]:
trabajadores = {'nombre':['Manuel','Maria','Diego','Bernarda','Alejandro','Felipe','Concha'],
           'antiguedad': [14, 26, 1, 2, 0, 6, 50],
           'profesion':['administrativo','director','becario','analista de datos','desempleado','abogado','retirada'],
           'salario':[1000, 2000, 100, 1000, 0, 1500, 800]}

trabajadores_df = pd.DataFrame(trabajadores)
trabajadores_df

nombre  antiguedad          profesion  salario
0     Manuel          14     administrativo     1000
1      Maria          26           director     2000
2      Diego           1            becario      100
3   Bernarda           2  analista de datos     1000
4  Alejandro           0        desempleado        0
5     Felipe           6            abogado     1500
6     Concha          50           retirada      800

#### merge

`merge` join datasets `on` one or more variables.
Another very important parameter to input is `how` we want to make the joint:
 - inner
 - right
 - left
 - outer



In [83]:
df_merge = familia_df.merge(trabajadores_df, on = 'nombre', how = 'left')
df_merge

miembro    nombre  edad estado_civil  altura    sexo  hombre  antiguedad  \
0    padre    Manuel    58       casado    1.78  hombro    True        14.0   
1    madre     Maria    56       casado    1.69   mujer   False        26.0   
2     hijo     Diego    17      soltero    1.70  hombre    True         1.0   
3     hija  Bernarda    23      soltero    1.79   mujer   False         2.0   
4  sobrino     Pablo     5      soltero    1.56  hombre    True         NaN   
5    primo   Antonio    12      soltero    1.59  hombre    True         NaN   
6   abuela    Concha    78        viuda    1.60   mujer   False        50.0   

           profesion  salario  
0     administrativo   1000.0  
1           director   2000.0  
2            becario    100.0  
3  analista de datos   1000.0  
4                NaN      NaN  
5                NaN      NaN  
6           retirada    800.0

In [84]:
df_merge = familia_df.merge(trabajadores_df, on = 'nombre', how = 'right')
df_merge

miembro     nombre  edad estado_civil  altura    sexo hombre  antiguedad  \
0   padre     Manuel  58.0       casado    1.78  hombro   True          14   
1   madre      Maria  56.0       casado    1.69   mujer  False          26   
2    hijo      Diego  17.0      soltero    1.70  hombre   True           1   
3    hija   Bernarda  23.0      soltero    1.79   mujer  False           2   
4     NaN  Alejandro   NaN          NaN     NaN     NaN    NaN           0   
5     NaN     Felipe   NaN          NaN     NaN     NaN    NaN           6   
6  abuela     Concha  78.0        viuda    1.60   mujer  False          50   

           profesion  salario  
0     administrativo     1000  
1           director     2000  
2            becario      100  
3  analista de datos     1000  
4        desempleado        0  
5            abogado     1500  
6           retirada      800

In [85]:
df_merge = familia_df.merge(trabajadores_df, on = 'nombre', how = 'inner')
df_merge

miembro    nombre  edad estado_civil  altura    sexo  hombre  antiguedad  \
0   padre    Manuel    58       casado    1.78  hombro    True          14   
1   madre     Maria    56       casado    1.69   mujer   False          26   
2    hijo     Diego    17      soltero    1.70  hombre    True           1   
3    hija  Bernarda    23      soltero    1.79   mujer   False           2   
4  abuela    Concha    78        viuda    1.60   mujer   False          50   

           profesion  salario  
0     administrativo     1000  
1           director     2000  
2            becario      100  
3  analista de datos     1000  
4           retirada      800

#### concat

`concat` is more flexible, you can join either by column or row input `axis` option equal to 1 for columns and 0 for rows

One advantage of `concat` face to `merge` is that you can join several tables at once

In [86]:
df_concat0 = pd.concat([familia_df, trabajadores_df], axis = 0)
df_concat0

miembro     nombre  edad estado_civil  altura    sexo hombre  antiguedad  \
0    padre     Manuel  58.0       casado    1.78  hombro   True         NaN   
1    madre      Maria  56.0       casado    1.69   mujer  False         NaN   
2     hijo      Diego  17.0      soltero    1.70  hombre   True         NaN   
3     hija   Bernarda  23.0      soltero    1.79   mujer  False         NaN   
4  sobrino      Pablo   5.0      soltero    1.56  hombre   True         NaN   
5    primo    Antonio  12.0      soltero    1.59  hombre   True         NaN   
6   abuela     Concha  78.0        viuda    1.60   mujer  False         NaN   
0      NaN     Manuel   NaN          NaN     NaN     NaN    NaN        14.0   
1      NaN      Maria   NaN          NaN     NaN     NaN    NaN        26.0   
2      NaN      Diego   NaN          NaN     NaN     NaN    NaN         1.0   
3      NaN   Bernarda   NaN          NaN     NaN     NaN    NaN         2.0   
4      NaN  Alejandro   NaN          NaN     NaN     NaN    NaN         0.0   
5      NaN     Felipe   NaN          NaN     NaN     NaN    NaN         6.0   
6      NaN     Concha   NaN          NaN     NaN     NaN    NaN        50.0   

           profesion  salario  
0                NaN      NaN  
1                NaN      NaN  
2                NaN      NaN  
3                NaN      NaN  
4                NaN      NaN  
5                NaN      NaN  
6                NaN      NaN  
0     administrativo   1000.0  
1           director   2000.0  
2            becario    100.0  
3  analista de datos   1000.0  
4        desempleado      0.0  
5            abogado   1500.0  
6           retirada    800.0

In [87]:
df_concat1 = pd.concat([familia_df, trabajadores_df], axis = 1)
df_concat1

miembro    nombre  edad estado_civil  altura    sexo  hombre     nombre  \
0    padre    Manuel    58       casado    1.78  hombro    True     Manuel   
1    madre     Maria    56       casado    1.69   mujer   False      Maria   
2     hijo     Diego    17      soltero    1.70  hombre    True      Diego   
3     hija  Bernarda    23      soltero    1.79   mujer   False   Bernarda   
4  sobrino     Pablo     5      soltero    1.56  hombre    True  Alejandro   
5    primo   Antonio    12      soltero    1.59  hombre    True     Felipe   
6   abuela    Concha    78        viuda    1.60   mujer   False     Concha   

   antiguedad          profesion  salario  
0          14     administrativo     1000  
1          26           director     2000  
2           1            becario      100  
3           2  analista de datos     1000  
4           0        desempleado        0  
5           6            abogado     1500  
6          50           retirada      800

### Exercise 1

Create new dataframes containing:
 - only Counties from Washington, Arizona, Utah and Montana States
 - only Counties with a TotalPop between 70000 and 100000 OR from Washington, Arizona, Utah and Montana States
 - the States with a Women population lower than 30000 AND Employed higher than 70000



In [60]:
# Exercise 1.1:
df_1_1 = df[(df['County'] == 'Washington') | (df['County'] == 'Arizona') | (df['County'] == 'Utah') | (df['County'] == 'Montana')]
len(df_1_1)

32

In [67]:
df_1_1.head(2)

CensusId     State      County  TotalPop  Hombres  Mujeres  Hispanic  \
64       1129   Alabama  Washington     16997     8490     8507       1.3   
182      5143  Arkansas  Washington    216432   108144   108288      16.0   

     White  Black  Native  ...  OtherTransp  WorkAtHome  MeanCommute  \
64    65.2   24.9     8.1  ...          1.0         1.3         32.6   
182   72.5    3.2     0.9  ...          1.3         4.6         21.3   

     Employed  PrivateWork  PublicWork  SelfEmployed  FamilyWork  \
64       5656         77.4        16.1           6.0         0.4   
182    102944         78.7        14.7           6.3         0.3   

     Unemployment  Unemployment_rate  
64           20.0               High  
182           6.2                Low  

[2 rows x 38 columns]

We check it:

In [62]:
(df['County'] == 'Washington').value_counts()

False    3189
True       31
Name: County, dtype: int64

In [63]:
(df['County'] == 'Utah').value_counts()

False    3219
True        1
Name: County, dtype: int64

In [64]:
(df['County'] == 'Arizona').value_counts()

False    3220
Name: County, dtype: int64

In [65]:
(df['County'] == 'Montana').value_counts()

False    3220
Name: County, dtype: int64

In [70]:
# Exercise 1.2: TotalPop between 70000 and 100000 OR from Washington, Arizona, Utah and Montana States
df_1_2 = df[(df['County'] == 'Washington') | (df['County'] == 'Arizona') | (df['County'] == 'Utah') | (df['County'] == 'Montana') | ((df['TotalPop'] > 70000) & (df['TotalPop'] < 100000))	]
len(df_1_2)

211

In [74]:
# Exercise 1.3: the States with a Women population lower than 30000 AND Employed higher than 70000
df_1_3 = df.loc[(df['Mujeres'] < 3000) & (df['Employed'] > 70000)]
df_1_3

Empty DataFrame
Columns: [CensusId, State, County, TotalPop, Hombres, Mujeres, Hispanic, White, Black, Native, Asian, Pacific, Citizen, Income, IncomeErr, IncomePerCap, IncomePerCapErr, Poverty, ChildPoverty, Professional, Service, Office, Construction, Production, Drive, Carpool, Transit, Walk, OtherTransp, WorkAtHome, MeanCommute, Employed, PrivateWork, PublicWork, SelfEmployed, FamilyWork, Unemployment, Unemployment_rate]
Index: []

[0 rows x 38 columns]

We proof that there isn't any row that matches with the required specifications

In [75]:
df[(df['Employed'] > 70000)].sort_values('Mujeres')

CensusId           State       County  TotalPop  Hombres  Mujeres  \
3113     55131       Wisconsin   Washington    132921    65880    67041   
1383     27139       Minnesota        Scott    137322    68449    68873   
840      19103            Iowa      Johnson    139436    69350    70086   
1957     37135  North Carolina       Orange    138644    66199    72445   
1792     34037      New Jersey       Sussex    145930    72391    73539   
...        ...             ...          ...       ...      ...      ...   
222       6073      California    San Diego   3223096  1618945  1604151   
103       4013         Arizona     Maricopa   4018143  1986158  2031985   
2623     48201           Texas       Harris   4356362  2166727  2189635   
610      17031        Illinois         Cook   5236393  2537245  2699148   
204       6037      California  Los Angeles  10038388  4945351  5093037   

      Hispanic  White  Black  Native  ...  OtherTransp  WorkAtHome  \
3113       2.9   93.6    1.0     0.1  ...          1.4         3.2   
1383       4.7   83.3    2.8     0.6  ...          1.1         5.0   
840        5.2   81.1    5.4     0.2  ...          4.1         4.1   
1957       8.3   69.8   11.5     0.4  ...          3.3         7.6   
1792       7.1   87.7    1.9     0.0  ...          1.0         5.4   
...        ...    ...    ...     ...  ...          ...         ...   
222       32.9   47.0    4.7     0.4  ...          2.0         6.8   
103       30.1   57.3    4.9     1.6  ...          2.6         5.9   
2623      41.6   31.7   18.5     0.2  ...          2.0         3.3   
610       24.7   43.1   23.7     0.1  ...          2.2         4.2   
204       48.2   26.9    8.0     0.2  ...          2.3         5.1   

      MeanCommute  Employed  PrivateWork  PublicWork  SelfEmployed  \
3113         24.7     72365         87.7         8.2           4.1   
1383         26.4     74498         86.6         9.2           4.1   
840          18.3     78838         67.0        28.0           4.8   
1957         22.3     70264         67.7        25.5           6.6   
1792         38.1     75669         78.8        14.3           6.7   
...           ...       ...          ...         ...           ...   
222          25.0   1462130         77.3        14.8           7.7   
103          25.5   1821038         82.5        11.7           5.7   
2623         28.2   2081889         83.4        10.1           6.3   
610          32.3   2463655         83.9        11.5           4.5   
204          30.0   4635465         79.0        11.5           9.4   

      FamilyWork  Unemployment  Unemployment_rate  
3113         0.1           4.5                Low  
1383         0.1           4.6                Low  
840          0.2           3.5                Low  
1957         0.2           6.4                Low  
1792         0.1           8.2                Low  
...          ...           ...                ...  
222          0.2           8.7                Low  
103          0.2           7.7                Low  
2623         0.1           7.5                Low  
610          0.1          10.7               High  
204          0.2          10.0                Low  

[413 rows x 38 columns]

As we can see, when we sort the values of the `Woman` when the `Employees`are more than `70000` the first row of `Woman` is 67041. COnsequently, this means that not exits any value for the conditions required.

### Exercise 2

Create a new dataframe containing:
 - only with estado_civil casado and salario >500.
 - Translate/Rename estado_civil and salario to English
 - Keep only these two columns

In [88]:
df_merge.columns

Index(['miembro', 'nombre', 'edad', 'estado_civil', 'altura', 'sexo', 'hombre',
       'antiguedad', 'profesion', 'salario'],
      dtype='object')

In [89]:
df_2_1 = df_merge.loc[(df_merge['estado_civil'] == 'casado') & (df_merge['salario'] >= 500)]

In [90]:
df_2_1

miembro  nombre  edad estado_civil  altura    sexo  hombre  antiguedad  \
0   padre  Manuel    58       casado    1.78  hombro    True          14   
1   madre   Maria    56       casado    1.69   mujer   False          26   

        profesion  salario  
0  administrativo     1000  
1        director     2000

In [91]:
df_2_2 = df_2_1.rename(columns={'estado_civil': 'civil_status', 'salario' : 'salary'})

In [92]:
df_2_2

miembro  nombre  edad civil_status  altura    sexo  hombre  antiguedad  \
0   padre  Manuel    58       casado    1.78  hombro    True          14   
1   madre   Maria    56       casado    1.69   mujer   False          26   

        profesion  salary  
0  administrativo    1000  
1        director    2000

In [93]:
df_2_3 = df_2_2[['civil_status', 'salary']]

In [94]:
df_2_3

civil_status  salary
0       casado    1000
1       casado    2000

### Exercise 3

 - Import stocks data (stocks.csv and company.csv)
 - Join these two tables by the field and union type(inner,left...) you consider more appropriate
 - Export in a csv the stock information of MSFT and ALK from 15 April 2021 to 31 December 2021, inlcuding Sector and Sub-industry information



In [95]:
df_stocks = pd.read_csv('/content/drive/MyDrive/USC Upstate/IronHack/Prework/DATA/data/stocks.csv', sep = ';')

In [96]:
df_stocks.head()

id        Date   Open   High    Low  Close   Volume ccode
0   0  2020-12-31  167.1  167.9  166.2  167.8  1841300   MMM
1   1  2021-01-04  168.0  169.1  163.7  165.0  2996200   MMM
2   2  2021-01-05  165.1  166.3  163.8  164.7  2295300   MMM
3   3  2021-01-06  165.8  168.5  165.2  167.2  3346400   MMM
4   4  2021-01-07  164.7  166.5  159.5  162.9  5863400   MMM

In [97]:
len(df_stocks)

153451

In [98]:
df_company = pd.read_csv('/content/drive/MyDrive/USC Upstate/IronHack/Prework/DATA/data/company.csv', sep = ';')

In [99]:
df_company.head()

Symbol     Security SEC filings  GICS Sector         GICS Sub-Industry  \
0    MMM           3M     reports  Industrials  Industrial Conglomerates   
1    AOS  A. O. Smith     reports  Industrials         Building Products   
2    ABT       Abbott     reports  Health Care     Health Care Equipment   
3   ABBV       AbbVie     reports  Health Care           Pharmaceuticals   
4   ABMD      Abiomed     reports  Health Care     Health Care Equipment   

     Headquarters Location Date first added        CIK      Founded  
0    Saint Paul, Minnesota       1976-08-09    66740.0         1902  
1     Milwaukee, Wisconsin       2017-07-26    91142.0         1916  
2  North Chicago, Illinois       1964-03-31     1800.0         1888  
3  North Chicago, Illinois       2012-12-31  1551152.0  2013 (1888)  
4   Danvers, Massachusetts       2018-05-31   815094.0         1981

In [100]:
len(df_company)

505

*   Inner Join: Only common rows.
*   Right Join: All rows from the right dataset and common rows from the left dataset.
*   Left Join: All rows from the left dataset and common rows from the right dataset.
*   Outer Join: All rows from both datasets, with nulls where there are no matches.

In [101]:
df_merge = pd.merge(df_stocks, df_company, left_on='ccode', right_on='Symbol', how='inner')


In [102]:
df_merge.head(3)

id        Date   Open   High    Low  Close   Volume ccode Symbol Security  \
0   0  2020-12-31  167.1  167.9  166.2  167.8  1841300   MMM    MMM       3M   
1   1  2021-01-04  168.0  169.1  163.7  165.0  2996200   MMM    MMM       3M   
2   2  2021-01-05  165.1  166.3  163.8  164.7  2295300   MMM    MMM       3M   

  SEC filings  GICS Sector         GICS Sub-Industry  Headquarters Location  \
0     reports  Industrials  Industrial Conglomerates  Saint Paul, Minnesota   
1     reports  Industrials  Industrial Conglomerates  Saint Paul, Minnesota   
2     reports  Industrials  Industrial Conglomerates  Saint Paul, Minnesota   

  Date first added      CIK Founded  
0       1976-08-09  66740.0    1902  
1       1976-08-09  66740.0    1902  
2       1976-08-09  66740.0    1902

In [105]:
# Filter the DataFrame to select rows for MSFT or ALK between April 15, 2021, and December 31, 2021
df_3_3 = df_merge.loc[(df_merge['Date'] >= '2021-04-15') & (df_merge['Date'] <= '2021-12-31') & ((df_merge['Symbol'] == 'MSFT') | (df_merge['Symbol'] == 'ALK'))]

# Export the DataFrame to a CSV file
df_3_3.to_csv('stock_3_3.csv', index=False)
